### ISSUE TITLE
Hi,I have a question about the 'yolo_layer.py'?

### ISSUE THREAD

**Issue by BCWang93** [DeNA/PyTorch_YOLOv3#31]
In the 'yolo_layer.py', the loss of the obj_conf have some error probability?
       ' loss_obj = self.bce_loss(output[..., 4], target[..., 4])'
this line in the code should have two part.According to the paper , I think this should have two parts is have obj and no_obj.I am not sure this,but I have some doubt,I think this should have two parts.Thanks!
Btw,this line '                    obj_mask[b, a, j, i] = 1' should be zero?Hope your reply!Thanks!

**Comment by hirotomusiker** [comment]
At the following line, the target tensor is set to zero (no-object).
https://github.com/DeNA/PyTorch_YOLOv3/blob/master/models/yolo_layer.py#L116-L117

Then at the following line, the target obj elements where ground-truth objects exist are set to 1. 
https://github.com/DeNA/PyTorch_YOLOv3/blob/master/models/yolo_layer.py#L174

Basically, the elements of obj_mask are one except the one where prediction has the larger IoU with ground-truth than ignore_thre.
https://github.com/DeNA/PyTorch_YOLOv3/blob/master/models/yolo_layer.py#L155

Finally by calculating
```
loss_obj = self.bce_loss(output[..., 4], target[..., 4])
```
, the elements of the obj channel are trained so that obj score becomes 1 where there is an object and zero elsewhere.

Thank you. I am planning to post the tutorial about target assignment to Medium (as Part 2).
https://medium.com/@hirotoschwert/reproducing-training-performance-of-yolov3-in-pytorch-part1-620140ad71d3

**Comment by BCWang93** [comment]
> At the following line, the target tensor is set to zero (no-object).
> https://github.com/DeNA/PyTorch_YOLOv3/blob/master/models/yolo_layer.py#L116-L117
> 
> Then at the following line, the target obj elements where ground-truth objects exist are set to 1.
> https://github.com/DeNA/PyTorch_YOLOv3/blob/master/models/yolo_layer.py#L174
> 
> Basically, the elements of obj_mask are one except the one where prediction has the larger IoU with ground-truth than ignore_thre.
> https://github.com/DeNA/PyTorch_YOLOv3/blob/master/models/yolo_layer.py#L155
> 
> Finally by calculating
> 
> ```
> loss_obj = self.bce_loss(output[..., 4], target[..., 4])
> ```
> , the elements of the obj channel are trained so that obj score becomes 1 where there is an object and zero elsewhere.
> 
> Thank you. I am planning to post the tutorial about target assignment to Medium (as Part 2).
> https://medium.com/@hirotoschwert/reproducing-training-performance-of-yolov3-in-pytorch-part1-620140ad71d3

Thanks for your reply!I understand it your means!But my means is I think it has two parts consists of the loss_conf like this:
'loss_conf=self.bce_loss(conf*mask, tconf*mask) + self.bce_loss(conf*noobj_mask,tconf*noobj_mask)'
because in other implement I see ,its almost like this.So I think in your implement, it should have two parts like this.
Btw,I want to ask,I trained the model in vocdata,but,the loss of  'x,y,w,h' don't  convergence,and the detection result is none use the trained model in the vocdata.But when I set the conf_thres=0.001,it has result in the detection.This problem confused me many days!Do you know this is why?Thank you very much!

**Comment by BCWang93** [comment]
> At the following line, the target tensor is set to zero (no-object).
> https://github.com/DeNA/PyTorch_YOLOv3/blob/master/models/yolo_layer.py#L116-L117
> 
> Then at the following line, the target obj elements where ground-truth objects exist are set to 1.
> https://github.com/DeNA/PyTorch_YOLOv3/blob/master/models/yolo_layer.py#L174
> 
> Basically, the elements of obj_mask are one except the one where prediction has the larger IoU with ground-truth than ignore_thre.
> https://github.com/DeNA/PyTorch_YOLOv3/blob/master/models/yolo_layer.py#L155
> 
> Finally by calculating
> 
> ```
> loss_obj = self.bce_loss(output[..., 4], target[..., 4])
> ```
> , the elements of the obj channel are trained so that obj score becomes 1 where there is an object and zero elsewhere.
> 
> Thank you. I am planning to post the tutorial about target assignment to Medium (as Part 2).
> https://medium.com/@hirotoschwert/reproducing-training-performance-of-yolov3-in-pytorch-part1-620140ad71d3

Btw,I have another problem.In other implement,I see the best_iou use the pred_box and truth_box to calculate to set the ignored box,and use the truth_box and reference anchors to calculate the iou used to mask the best_n.In your implement,I see you use the reference anchors and truth_box iou to mask.So,do you  think there two ways are better?Thanks!

**Comment by BCWang93** [comment]
> At the following line, the target tensor is set to zero (no-object).
> https://github.com/DeNA/PyTorch_YOLOv3/blob/master/models/yolo_layer.py#L116-L117
> 
> Then at the following line, the target obj elements where ground-truth objects exist are set to 1.
> https://github.com/DeNA/PyTorch_YOLOv3/blob/master/models/yolo_layer.py#L174
> 
> Basically, the elements of obj_mask are one except the one where prediction has the larger IoU with ground-truth than ignore_thre.
> https://github.com/DeNA/PyTorch_YOLOv3/blob/master/models/yolo_layer.py#L155
> 
> Finally by calculating
> 
> ```
> loss_obj = self.bce_loss(output[..., 4], target[..., 4])
> ```
> , the elements of the obj channel are trained so that obj score becomes 1 where there is an object and zero elsewhere.
> 
> Thank you. I am planning to post the tutorial about target assignment to Medium (as Part 2).
> https://medium.com/@hirotoschwert/reproducing-training-performance-of-yolov3-in-pytorch-part1-620140ad71d3

In the  other hand,do you have the trained loss plot?Can you increase it to the implement?Thanks!

**Comment by hirotomusiker** [comment]
First of all, our implementation is based on **the original implementation of the yolo_layer** to reproduce the training results exactly. 
https://github.com/pjreddie/darknet/blob/master/src/yolo_layer.c

> it should have two parts like this.

No. We don't need to divide the obj loss into obj and no_obj in YOLOv3 anymore. The weights for the two losses are the same, unlike YOLOv2.

> 'x,y,w,h' don't convergence when using VOC

Please check the bounding box format. We haven't tested our implementation as for VOC.

> pred_box and truth_box and reference anchors

We take the same procedure as the original darknet's yolo_layer. 

> increase the trained loss plot

I cannot get what you mean but we don't have a plan to upload a loss plot right now. 

Thank you.

**Comment by BCWang93** [comment]
> First of all, our implementation is based on **the original implementation of the yolo_layer** to reproduce the training results exactly.
> https://github.com/pjreddie/darknet/blob/master/src/yolo_layer.c
> 
> > it should have two parts like this.
> 
> No. We don't need to divide the obj loss into obj and no_obj in YOLOv3 anymore. The weights for the two losses are the same, unlike YOLOv2.
> 
> > 'x,y,w,h' don't convergence when using VOC
> 
> Please check the bounding box format. We haven't tested our implementation as for VOC.
> 
> > pred_box and truth_box and reference anchors
> 
> We take the same procedure as the original darknet's yolo_layer.
> 
> > increase the trained loss plot
> 
> I cannot get what you mean but we don't have a plan to upload a loss plot right now.
> 
> Thank you.

Thanks for your reply!I will check it and use your ways to train the model again!If possible,I will post the result in this issu.Thank you very much!

**Comment by eleflea** [comment]
@BCWang93 I have trained on VOC dataset too, and get a poor mAP. Have you solve this problem?
